design/ingressroute-design.md

```diff
@@ -0,0 +1,300 @@
+  virtualhost:
+    # the fully qualified domain name of the root of the ingress tree
+    # all leaves of the DAG rooted at this object relate to the fqdn (and its aliases)
+    fqdn: www.google.com
+    # a set of aliases for the domain, these may be alternative fqdns which are considered
+    # aliases of the primary fqdn
+    aliases:
+      - www.google.com.au
+      - google.com
+    # if present describes tls properties. The CNI names that will be matched on
+    # are described in fqdn and aliases, the tls.secretName secret must contain a
+    # matching certificate
+    tls:
+      # required, the name of a secret in the current namespace
+      secretName: google-tls
+      # other properties like cipher suites may be added later
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific.
+  routes:
+  # each route entry starts with a prefix match
+  # and one of service or delegate
+  - match: /static
+    # service defines the properties of the service in the current namespace
+    # that will handle traffic matching the route
+    service:
+    - name: google-static
+      port: 9000
+  - match: /finance
+    # delegate delegates the matching route to another IngressRoute object.
+    # This delegates the responsibility for /finance to the IngressRoute matching the delegate parameters
+    delegate:
+      # the name of the IngressRoute object to delegate to
+      name: google-finance
+      # the namespace of the IngressRoute object, if blank the current namespace is assumed
+      namespace: finance
+  - match: /ads
+    # more than one service name/port pair may be present allowing the services for a route to
+    # be handled by more than one service.
+    service:
+    - name: ads-red
+      port: 9090
+    - name: ads-blue
+      port: 9090
+```
+
+This is an example of the `google-finance` object which has been delegated responsibilty for paths starting with `/finance`.
+
+```yaml
+apiVersion: contour.heptio.com/v1alpha1
+kind: IngressRoute
+metadata:
+  name: google-finance
+  namespace: finance
+spec:
+  # note that this is an edge, so there is no virtualhost key
+  # routes contains the set of routes for this virtual host.
+  # routes must _always_ be present and non empty.
+  # routes can be present in any order, and will be matched from most to least
+  # specific, however as this is an edge, only prefixes that match the prefix
+  # that delegated to this object.
+  routes:
+  # each route in this object must start with /finance as this was the prefix which
+  # delegated to this object.
+  - match: /finance
+    service:
+    - name: finance-app
+      port: 9999
+  - match: /finance/static
+    service:
+    - name: finance-static-content
+      port: 8080
+  - match: /finance/partners
+    # delegate the /finance/partners prefix to the partners namespace
+    delegate:
```

Comment:
The only thing that could be troublesome for users is now they've got to know what names are defined to be delegated to from the root. So I I'm a user in a namespace and my Admin tells me that I've got `/` for `foo.com`, there's a coordination that needs to happen to match up the names since I (most likely) can't see the root object.

Reply:
I hoped that this information could be reflected in the delegationStatus mentioned below; but it's only going to be updated once you get everything linked up correctly.

I don't really have an answer for that one, but do see the update i'm pushing the security concerns section which might address this.

Reply:
Other options are to have a CLI tool scaffold this out for users or have Contour scaffold in the delegated namespace an empty route which includes all the items available. It will be invalid until the user adds services, but at least helps this out.

Reply:
@jbeda i know you want delegated routes to be relative to the thing they are delegated from, but that makes understanding where the route is being served from contingent on knowing the path of delegation that lead to your IngressRoute CRD; there may be multiple paths, even within a single root. I'm not closing the door on relative routes, the syntax for them could be as simple as not including the "/" prefix; just like we do with open(2), but I don't want to lead with it, as it's bloody complex.